src/main/java/edu/harvard/iq/dataverse/api/Admin.java

```diff
@@ -1234,4 +1244,135 @@ public Response clearMetricsCacheByName(@PathParam("name") String name) {
 	}
 
+    @GET
+    @Path("/dataverse/{alias}/addAdminsToChildren")
+    public Response addAdminsToChildren(@PathParam("alias") String alias) throws WrappedResponse {
+        Dataverse owner = dataverseSvc.findByAlias(alias);
+        if (owner == null) {
+            return error(Response.Status.NOT_FOUND,
+                    "Could not find dataverse based on alias supplied: " + alias + ".");
+        }
+        try {
+            AuthenticatedUser user = findAuthenticatedUserOrDie();
+            if (!user.isSuperuser()) {
+                return error(Response.Status.FORBIDDEN, "Superusers only.");
+            }
+        } catch (WrappedResponse wr) {
+            return wr.getResponse();
+        }
+        if (settingsSvc.isTrueForKey(SettingsServiceBean.Key.InheritParentAdmins, false)) {
+
+            String qstr = "WITH RECURSIVE path_elements AS ((" + " SELECT id, dtype FROM dvobject WHERE id in ("
+                    + owner.getId() + "))" + " UNION\n"
+                    + " SELECT o.id, o.dtype FROM path_elements p, dvobject o WHERE o.owner_id = p.id and o.dtype='Dataverse') "
+                    + "SELECT id FROM path_elements WHERE id !=" + owner.getId() + ";"; // ORDER by id ASC;";
+
+            List<Integer> childIds;
+
+            try {
+                childIds = em.createNativeQuery(qstr).getResultList();
+            } catch (Exception ex) {
+                childIds = null;
+            }
+
+            if (childIds == null || childIds.size() < 1) {
+                return error(Response.Status.NOT_FOUND,
+                        "Could not find any child dataverses based on alias supplied: " + alias + ".");
+            }
+            JsonArrayBuilder usedNames = Json.createArrayBuilder();
+            JsonArrayBuilder unusedNames = Json.createArrayBuilder();
+            JsonArrayBuilder dataverseIds = Json.createArrayBuilder();
+            JsonArrayBuilder dataverseAliases = Json.createArrayBuilder();
+            try {
+                List<Dataverse> children = new ArrayList<Dataverse>();
+
+                for (int i=0; i < childIds.size(); i++) {
+                    Integer childId = childIds.get(i);
+                    Dataverse child = dataverseSvc.find(new Long(childId.longValue()));
+                    if (child != null) {
+                        children.add(child);
+                        dataverseIds.add(childId.longValue());
+                        dataverseAliases.add(child.getAlias());
+                    }
+                }
+
+                List<RoleAssignment> assignedRoles = rolesSvc.directRoleAssignments(owner);
+                // Find the built in admin role (currently by alias)
+                DataverseRole adminRole = rolesSvc.findBuiltinRoleByAlias(DataverseRole.ADMIN);
+                String privateUrlToken = null;
+
+                Map<Long, List<RoleAssignment>> existingRAs = new HashMap<Long, List<RoleAssignment>>();
+                for (Dataverse childDv : children) {
```

Comment:
This loop will be very inefficient as it will be computing a ton of redundant permissions information.

Reply:
@oscardssmith - I'm not sure I'm following this and the other comment below about using a map. The one way I saw to improve efficiency was to filter the list of roles on the owner to find just the admin roles and to use that smaller list in the subsequent loops. That's been added to the PR.